packages/api/lib/ems.js

```diff
@@ -0,0 +1,280 @@
+  let key = originalKey;
+  let exists = await aws.fileExists(bucket, key);
+  let i = 1;
+  while (exists) {
+    key = `${originalKey}.rev${i}`;
+    exists = await aws.fileExists(bucket, key); // eslint-disable-line no-await-in-loop
+    i += 1;
+  }
+
+  await aws.s3().putObject({
+    Bucket: bucket,
+    Key: key,
+    Body: fs.createReadStream(filename)
+  }).promise();
+
+  fs.unlinkSync(filename);
+  const s3Uri = `s3://${bucket}/${key}`;
+  log.info(`uploaded ${s3Uri}`);
+  return s3Uri;
+}
+
+/**
+ * build report file name
+ * The report filename is in format:
+ * <YYYYMMDD> _<Provider>_<FileType>_<DataSource>.flt.rev<1-n>
+ *
+ * @param {string} reportType - report type (ingest, archive, delete)
+ * @returns {string} - report file name
+ */
+function buildReportFileName(reportType) {
+  // DataSource: designates the database table name or data source file/table name
+  // use stackname as DataSource for now
+  const provider = process.env.ems_provider || 'cumulus';
+  const dataSource = process.env.stackName;
+  const datestring = moment.utc().format('YYYYMMDD');
+  let fileType = 'Ing';
+  if (reportType === 'archive') fileType = 'Arch';
+  else if (reportType === 'delete') fileType = 'ArchDel';
+  return `${datestring}_${provider}_${fileType}_${dataSource}.flt`;
+}
+
+/**
+ * generate an EMS report
+ *
+ * @param {string} reportType - report type (ingest, archive, delete)
+ * @param {string} startTime - start time of the records
+ * @param {string} endTime - end time of the records
+ * @returns {Object} - report type and its file path {reportType, file}
+ */
+async function generateReport(reportType, startTime, endTime) {
+  log.debug(`ems.generateReport ${reportType} startTime: ${startTime} endTime: ${endTime}`);
+
+  // create a temporary file for the report
+  const name = buildReportFileName(reportType);
+  const filename = path.join(os.tmpdir(), name);
+  const stream = fs.createWriteStream(filename);
+
+  // retrieve granule/deletedgranule records in batches, and generate EMS records for each batch
+  const esClient = await Search.es();
+  const type = (reportType !== 'delete') ? 'granule' : 'deletedgranule';
+
+  const esIndex = process.env.ES_INDEX || defaultIndexAlias;
+  const searchQuery = buildSearchQuery(esIndex, type, startTime, endTime);
+  let response = await esClient.search(searchQuery);
+  let granules = response.hits.hits.map((s) => s._source);
+  let numRetrieved = granules.length;
+  stream.write(buildEMSRecords(get(emsMappings, reportType), granules).join('\n'));
+
+  while (response.hits.total !== numRetrieved) {
+    response = await esClient.scroll({ // eslint-disable-line no-await-in-loop
+      scrollId: response._scroll_id,
+      scroll: '30s'
+    });
+    granules = response.hits.hits.map((s) => s._source);
+    stream.write('\n');
+    stream.write(buildEMSRecords(get(emsMappings, reportType), granules).join('\n'));
+    numRetrieved += granules.length;
```

Comment:
These few lines are identical to lines further up in the function.  Should probably be broken out into a separate function in that case.